File: AnotherBlogSite.Application/Services/BlogPostsService.cs
using AnotherBlogSite.Application.Entities;
using AnotherBlogSite.Common;
using AnotherBlogSite.Data;
using AnotherBlogSite.Data.Entities;
using Microsoft.EntityFrameworkCore;
using BlogPostMapper = AnotherBlogSite.Application.Mapper.BlogPostMapper;

namespace AnotherBlogSite.Application.Services;

internal sealed class BlogPostsService: IBlogPostService
{
    private readonly BlogPostMapper _mapper = new();
    private readonly BlogSiteContext _context;

    public BlogPostsService(BlogSiteContext context)
    {
        _context = context;
    }

    public Task<List<BlogPostModel>> GetAllAsync()
    {
        return _mapper.ProjectToModel(_context.BlogPosts.Include(x => x.Author).AsNoTracking()).ToListAsync();
    }

    public async Task<Result<BlogPostModel>> GetAsync(Guid blogPostId)
    {
        var blogPost = await _context.BlogPosts
            .AsNoTracking()
            .Include(x => x.Author)
            .Include(x => x.Comments)
                .ThenInclude(x => x.Author)
            .SingleOrDefaultAsync(x => x.Id == blogPostId);

        if (blogPost is null)
            return Result<BlogPostModel>.CreateFailure("Blog Post not found!", ErrorType.NotFound);

        return Result<BlogPostModel>.CreateSuccess(_mapper.MapWithComments(blogPost));
    }

    public async Task<Result<BlogPostModel>> CreateAsync(string title, string content, Guid authorId)
    {
        var newBlogPost = new BlogPost()
        {
            Title = title,
            AuthorId = authorId,
            Content = content,
            CreatedDate = DateTimeOffset.UtcNow,
        };

        try
        {
            _context.BlogPosts.Add(newBlogPost);

            int createdCount = await _context.SaveChangesAsync();

            if (createdCount != 1)
                return Result<BlogPostModel>.CreateFailure("Failed to create Blog Post!");

            await _context.Entry(newBlogPost).Reference(x => x.Author).LoadAsync();

            return Result<BlogPostModel>.CreateSuccess(_mapper.MapWithoutComments(newBlogPost));
        }
        catch (DbUpdateException)
        {
            return Result<BlogPostModel>.CreateFailure("Failed to update Blog Post!");
        }
    }

    public async Task<Result<BlogPostModel>> UpdateAsync(Guid blogPostId, string title, string content)
    {
        var oldBlogPost = await _context.BlogPosts
            .Include(x => x.Author)
            .Include(x => x.Comments)
                .ThenInclude(x => x.Author)
            .SingleOrDefaultAsync(x => x.Id == blogPostId);

        if (oldBlogPost is null)
            return Result<BlogPostModel>.CreateFailure("Blog Post not found!", ErrorType.NotFound);

        oldBlogPost.Title = title;
        oldBlogPost.Content = content;

        try
        {
            await _context.SaveChangesAsync();
        }
        catch (DbUpdateException)
        {
            return Result<BlogPostModel>.CreateFailure("Failed to update Comment!");
        }

        return Result<BlogPostModel>.CreateSuccess(_mapper.MapWithComments(oldBlogPost));
    }

    public Task DeleteAsync(Guid blogPostId)
    {
        return _context.BlogPosts.Where(x => x.Id == blogPostId).ExecuteDeleteAsync();
    }
}

File: AnotherBlogSite.Presentation/ClientApp/services/BlogPostsService.ts
﻿import IBlogPost from "../models/IBlogPost.ts";
import Guid from "../models/Guid.ts";
import IRequestProvider from "../models/IRequestProvider.ts";

export interface CreateBlogPost {
    title: string;
    content: string;
}

export interface UpdateBlogPost {
    blogPostId: Guid;
    title: string;
    content: string;
}

export default class BlogPostsService {
    #requestProvider: IRequestProvider;

    constructor(requestProvider: IRequestProvider) {
        this.#requestProvider = requestProvider;
    }

    getBlogPosts = async (): Promise<IBlogPost[]> => {
        const data = await this.#requestProvider.get<IBlogPost[]>('/BlogPosts');

        data.forEach(x => x.createdDate = new Date(x.createdDate));

        return data;
    }

    getBlogPost = async (blogPostId: Guid): Promise<IBlogPost> => {
        const data = await this.#requestProvider.get<IBlogPost>(`/BlogPosts/${blogPostId}`);

        data.createdDate = new Date(data.createdDate);
        data.comments?.forEach(comment => comment.createdDate = new Date(comment.createdDate));

        return data;
    }

    createBlogPost = async (createBlogPost: CreateBlogPost): Promise<IBlogPost> => {
        const data = await this.#requestProvider.post<CreateBlogPost, IBlogPost>('/BlogPosts', createBlogPost);

        data.createdDate = new Date(data.createdDate);
        data.comments?.forEach(comment => comment.createdDate = new Date(comment.createdDate));

        return data;
    }

    updateBlogPost = async (updateBlogPost: UpdateBlogPost): Promise<IBlogPost> => {
        const data = await this.#requestProvider.put<UpdateBlogPost, IBlogPost>('/BlogPosts', updateBlogPost);

        data.createdDate = new Date(data.createdDate);
        data.comments?.forEach(comment => comment.createdDate = new Date(comment.createdDate));

        return data;
    }

    deleteBlogPost = async (blogPostId: Guid): Promise<void> => {
        console.log(this.#requestProvider);
        await this.#requestProvider.delete<null>(`/BlogPosts/${blogPostId}`);
    }
}